{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "David Adams\n",
    "\n",
    "1.a. \n",
    "Data in Millikan:\n",
    "<div class=\"alert alert-info\">\n",
    "8/10\n",
    "I would say that you had some nice ideas, but you should spell out your thinking a little more to be clearer and you should use upper and lower bound estimates for some estimates.\n",
    "</div>\n",
    "\n",
    "\n",
    "    Area of 1 floor of Millikan = 3 * 10^2 m^2\n",
    "    * 9 floors = 2.7 * 10^3 m^2\n",
    "    * 1 bookshelf / 5 m^2 = 5.4 * 10^2 shelves\n",
    "    * 10^2 books / 1 shelf = 5.4 * 10^4 books\n",
    "    * 2 * 10^2 pages / 1 book = 10^7 pages\n",
    "    * 2.5 * 10^2 words / page = 2.5 * 10^9 words\n",
    "    * 5 characters / 1 word = 10^10 characters\n",
    "    * 1 byte / 1 character = 10^10 bytes = 10 gigabytes\n",
    "     \n",
    "Data in E. coli genome:\n",
    "\n",
    "    4.6 * 10^6 base pairs\n",
    "    * 2 bits / 1 base pair = 10^7 bits\n",
    "    * 1 byte / 8 bits = 10^6 bytes = 1 megabyte\n",
    "     \n",
    "Data passed from parent to offspring (Human):\n",
    "\n",
    "    6.5 *10^9 base pairs\n",
    "    / 2 (since half comes from each parent) = 3.2 * 10^9 base pairs\n",
    "    * 2 bits / 1 base pair = 6.5 *10^9 bits\n",
    "    * 1 byte / 8 bits = 8 * 10^8 bytes = 1 gigabyte\n",
    " \n",
    "1.b.\n",
    "Information density in influenza virus:\n",
    "<div class=\"alert alert-info\">\n",
    "9/9\n",
    "</div>\n",
    "\n",
    "\n",
    "    1.4 * 10^3 nucleotides\n",
    "    * 2 bits / 1 nucleotide = 2.8 * 10^3 bits\n",
    "    * 1 byte / 8 bits = 3.5 * 10^2 bytes\n",
    "    \n",
    "    100 nm diameter\n",
    "    / 2 = 50 nm radius\n",
    "    4/3 * pi * r^3 = 5 * 10^5 nm^3 volume\n",
    "    \n",
    "    3.5 * 10^2 bytes / 5 * 10^5 nm^3 = 7 * 10^-4 bytes/nm^3\n",
    "    * 10^21 nm^3 / 1 cm^3 = 7 * 10^17 bytes/cm^3\n",
    "    \n",
    "Information density in thumb drive:\n",
    "\n",
    "    data stored = 10 gigabytes\n",
    "    * 10^9 bytes /  1 gigabyte = 10^10 bytes\n",
    "    \n",
    "    volume = 10 cm^3 (rectangle with sides 5cm x 2 cm x 1cm)\n",
    "    \n",
    "    10^10 bytes/ 10 cm^3 = 10^9 bytes/cm^3\n",
    "    \n",
    "The information density of the influenza virus is much higher than that of a standard thumb drive.\n",
    "\n",
    "1.c.\n",
    "SRA to Harry Potter series:\n",
    "<div class=\"alert alert-info\">\n",
    "5/6\n",
    "Remember to explain the logic behind your estimates!\n",
    "</div>\n",
    "\n",
    "\n",
    "    bases in SRA = 9.8 * 10^15 bases\n",
    "    \n",
    "    books in Harry Potter series = 7 books\n",
    "    * 5 * 10^2 pages / 1 book = 3.5 * 10^3 pages\n",
    "    * 2.5 * 10^2 words / 1 page = 10^6 words\n",
    "    * 5 characters / 1 word = 5 * 10^6 characters\n",
    "    \n",
    "    9.8 * 10^15 bases / 5 * 10^6 characters = 2 * 10^9 copies of Harry Potter series\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "2a: 12/12\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'hw1_dataset/frogs_st.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-f8ae47d065bb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m \u001b[0;31m# Open the files for the Sao Tome and Africa sequences\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'hw1_dataset/frogs_st.txt'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mseq\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m     \u001b[0mst_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mBio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'fasta'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m     \u001b[0mst_sequences\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mrecord\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mrecord\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mst_file\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'hw1_dataset/frogs_st.txt'"
     ]
    }
   ],
   "source": [
    "# 2.a.\n",
    "\n",
    "# Imports\n",
    "import Bio.SeqIO\n",
    "\n",
    "def compare_sequences(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Compares 2 sequences of the same length and computes the percent similarity between them ignoring gaps.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    seq1, seq2 : string\n",
    "        DNA sequences, must be the same length\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    score : float\n",
    "        percent similarity between seq1 and seq2\n",
    "    \"\"\"\n",
    "    if len(seq1) != len(seq2):\n",
    "        raise ValueError('Sequences must be of the same length')\n",
    "        \n",
    "    seq1 = seq1.lower()\n",
    "    seq2 = seq2.lower()\n",
    "    \n",
    "    num_sim = 0\n",
    "    num_comp = 0\n",
    "    \n",
    "    for base in range(len(seq1)):\n",
    "        if seq1[base] != '-' and seq2[base] != '-':\n",
    "            if seq1[base] == seq2[base]:\n",
    "                num_sim += 1\n",
    "            num_comp += 1\n",
    "    \n",
    "    score = num_sim / num_comp\n",
    "    \n",
    "    return score\n",
    "\n",
    "# Open the files for the Sao Tome and Africa sequences\n",
    "with open('hw1_dataset/frogs_st.txt', 'r') as seq:\n",
    "    st_file = Bio.SeqIO.parse(seq, 'fasta')\n",
    "    st_sequences = [record for record in st_file]\n",
    "with open('hw1_dataset/frogs_africa.txt', 'r') as seq:\n",
    "    africa_file = Bio.SeqIO.parse(seq, 'fasta')\n",
    "    africa_sequences = [record for record in africa_file]\n",
    "\n",
    "# Compare each of the 3 frogs from Sao Tome to the frogs from Africa and print top 3 results\n",
    "c = 1\n",
    "for place1 in st_sequences:\n",
    "    # Initialize variables to hold best 3 matches and scores for them\n",
    "    match1 = africa_sequences[0]\n",
    "    match1_score = compare_sequences(place1.seq, match1.seq)\n",
    "    match2 = africa_sequences[1]\n",
    "    match2_score = compare_sequences(place1.seq, match2.seq)\n",
    "    match3 = africa_sequences[2]\n",
    "    match3_score = compare_sequences(place1.seq, match3.seq)\n",
    "    \n",
    "    # Sort our initial best 3 in order\n",
    "    if(match2_score > match1_score):\n",
    "        temp = match1\n",
    "        temp_score = match1_score\n",
    "        match1 = match2\n",
    "        match1_score = match2_score\n",
    "        match2 = temp\n",
    "        match2 = temp_score\n",
    "    if(match3_score > match1_score):\n",
    "        temp = match1\n",
    "        temp_score = match1_score\n",
    "        match1 = match3\n",
    "        match1_score = match3_score\n",
    "        match3 = temp\n",
    "        match3_score = temp_score\n",
    "    if(match3_score > match2_score):\n",
    "        temp = match2\n",
    "        temp_score = match2_score\n",
    "        match2 = match3\n",
    "        match2_score = match3_score\n",
    "        match3 = temp\n",
    "        match3_score = temp_score\n",
    "    \n",
    "    # Compare sequences and find best 3 matches\n",
    "    for x in range(3, len(africa_sequences)):\n",
    "        score = compare_sequences(place1.seq, africa_sequences[x].seq)\n",
    "        if(score > match1_score):\n",
    "            match1 = africa_sequences[x]\n",
    "            match1_score = score\n",
    "        elif(score > match2_score):\n",
    "            match2 = africa_sequences[x]\n",
    "            match2_score = score\n",
    "        elif(score > match3_score):\n",
    "            match3 = africa_sequences[x]\n",
    "            match3_score = score\n",
    "\n",
    "    # Print stuff\n",
    "    print('Comparing to frog ' + str(c) + ' from Sao Tome (' + place1.description + '):')\n",
    "    print('Top 3 Matches:')\n",
    "    print('1. ' + match1.description + ': Score = ' + str(match1_score))\n",
    "    print('2. ' + match2.description + ': Score = ' + str(match2_score))\n",
    "    print('3. ' + match3.description + ': Score = ' + str(match3_score))\n",
    "    print()\n",
    "    c += 1\n",
    "\n",
    "# 2 of the frogs in the data set (which happen to be the 2nd and 3rd best matches) give the same score\n",
    "# The 2nd and 3rd best matches are different frogs though"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2.b.\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "12/13 : Overall nicely done.\n",
    "</div>\n",
    "\n",
    "    Based on the DNA sequences the Ptychadena newtoni likely originated from Tanzania.\n",
    "    \n",
    "    The Congo River is likely the path that a frog on a raft would have to take to get from Tanzania to Sao Tome. It is fed by Lake\n",
    "    Tanganyika, which is on the western border of Tanzania. Let's say that the Congo River has serious enough flooding to create rafts\n",
    "    10 times a year. The Congo River is almost 3000 miles long. Since there is a very large number of trees along the banks of the river,\n",
    "    estimate that 1 raft forms for every 5 feet of river.\n",
    "    \n",
    "<div class=\"alert alert-info\">\n",
    "Why look within such short range of the river banks? Why not look farther? If you think about flash floods in the US, you can envision all sorts of debris washing through town. I can imagine that to be the case for the floods in Africa as well.\n",
    "</div>\n",
    "    \n",
    "    That gives us about 1000 rafts a mile or 3 * 10^6 rafts over the course of the\n",
    "    entire river.     \n",
    "    However since Tanzania is on the east coast of Africa, frogs are not living everywhere along the banks of the Congo\n",
    "    River. We will say that only 1/3 of the Congo River has the frog living on its banks, giving us 10^6 possible rafts. If the frog is\n",
    "    common, there should be about 1 group of frogs on every raft tree that ends up in the river. That gives us 10^6 groups of frogs that\n",
    "    could potentially survive the journey. Let's say that for every mile of the Congo River 1% of the groups die. Over 3000 miles that\n",
    "    gives us 8 * 10^-8 groups. In other words, we expect a group of frogs to survive the trip about once every 10^-7 floods. With 10 floods\n",
    "    a year over 13 * 10^6 years, we expect 10 groups of frogs to reach Sao Tome over 13 million years. It is possible then that a\n",
    "    population of frogs originating from Tanzania could have reached Sao Tome. \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3.a. \n",
    "\n",
    "![title](hw1_3a.png)\n",
    "    \n",
    "\n",
    "3.b.\n",
    "\n",
    "    Haiti strain was collected 11/01/10 while Nepal- Strain 14 was collected 8/01/10, so we will assume the mutations occurred in those 3\n",
    "    months.\n",
    "    2 SNPs / 3 months\n",
    "    * 1 month / 30 days = 1 SNP / 45 days\n",
    "    * 1 day / 1440 minutes = 1 SNP / 6.5 * 10^4 minutes\n",
    "    * 40 minutes / 1 division = 1 SNP / 1500 divisions\n",
    "    / 4 * 10^6 base pairs = 1 SNP / 6 * 10^9 base pairs per division\n",
    "    = 2 * 10^-10 SNPs per base pair per division\n",
    "    \n",
    "<div class=\"alert alert-info\">\n",
    "10/10\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.a.\n",
    "\n",
    "    First, the landmass that today forms New Zealand and New Caledonia split from the landmass that today forms Southern South America,\n",
    "    West Antarctica, Australia, and New Guinea about 80 MYA. Next, New Zealand and New Caledonia split apart about 70 MYA. Then about 50\n",
    "    MYA the landmass that today is Southern South America and West Antarctica split from what today is Australia and New Guinea. Then\n",
    "    about 30 MYA Southern South America and West Antarctica split. Finally, Australia and New Guinea split apart about 10 MYA.\n",
    "<div class=\"alert alert-info\">\n",
    "10/10\n",
    "</div>\n",
    "\n",
    "4.b.\n",
    "\n",
    "    New Zealand and Australia were last connected about 80 MYA\n",
    "    \n",
    "    7 mutations per billion base pairs per generation\n",
    "    * 1 generation / 20 years = .35 mutations per billion base pairs per year\n",
    "    * 500 * 10^6 base pairs / 1 genome = .18 mutations per year\n",
    "    * 80 * 10^6 years = 1.4 * 10^7 mutations\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "8/8\n",
    "</div>\n",
    " \n",
    "4.c.\n",
    "    \n",
    "    7 mutations per billion base pairs per generation\n",
    "    * 1 generation / 20 years = .35 mutations per billion base pairs per year\n",
    "    * 500 * 10^6 base pairs / 1 genome = .18 mutations per year\n",
    "    \n",
    "    9 * 10^6 mutations \n",
    "    * 1 year / .18 mutations = 5 * 10^7 years = 50 MYA\n",
    "    \n",
    "    The Nothofagus species likely diverged due to dispersal rather than vicariance. \n",
    "   \n",
    " <div class=\"alert alert-info\">\n",
    "7/7\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
